### PR TITLE
feat(common.opcua): Add option to include OPC-UA DataType as a Field

### LINKED ISSUE #8669: inputs.opcua: Add metric for datatype
## Feature Request
Multiple attributes to pass from opcua-input plugin 

### Proposal:
I'm new to Telegraf, So, Using opcua-input plugin, I only get 1 value from the identitfier which is its main value but I also want to monitor other attributes like its datatype with it

### Current behavior:
opcua,host=sysadmin-Virtual-Machine Tartget_Production_Achieved=true,Quality="OK (0x0)" 1610354410000000000
opcua,host=sysadmin-Virtual-Machine Tartget_Production_Count=0i,Quality="OK (0x0)" 1610354410000000000

### Desired behavior:
opcua,host=sysadmin-Virtual-Machine Tartget_Production_Achieved=true, datatype=boolean, Quality="OK (0x0)" 1610354410000000000
opcua,host=sysadmin-Virtual-Machine Tartget_Production_Count=0i, datatype=int16, Quality="OK (0x0)" 1610354410000000000

### Use case: 
I want to monitor multiple device and handling the data can be easier if I know what datatype it is.

### PR DESCRIPTION
**Pull request by caallinson**
- [x] Updated associated README.md.
- [x] Wrote appropriate unit tests.
- [x] Pull request title or commits are in [conventional commit format](https://www.conventionalcommits.org/en/v1.0.0/#summary)

resolves #8669

Adds a structure for optional fields to be included with metrics, and implements a method to add DataType as a field when OPC-UA tags are read

### CODE REVIEW

File: plugins/inputs/opcua/sample.conf [hunk 1]
```diff
@@ -123,6 +123,11 @@
   #   identifier_type = ""
   #   identifier = ""
 
+  ## Include additional Fields in each metric
+  # [inputs.opcua.additional_fields]
+    ## OPC-UA DataType
+    # include_datatype = true
```
Comment (Hipska) [inline plugins/inputs/opcua/sample.conf:129]: The commented out value should represent the default value
```suggestion
    # include_datatype = false
```

File: plugins/inputs/opcua_listener/sample.conf [hunk 2]
```diff
@@ -187,6 +187,12 @@
   #       deadband_type = "Absolute"
   #       deadband_value = 0.0
   #
+
+  ## Include additional Fields in each metric
+  # [inputs.opcua.additional_fields]
+    ## OPC-UA DataType
+    # include_datatype = true
```
Comment (Hipska) [inline plugins/inputs/opcua_listener/sample.conf:194]: The commented out value should represent the default value
```suggestion
    # include_datatype = false
```

File: plugins/common/opcua/client.go [hunk 3]
```diff
@@ -18,6 +18,10 @@ type OpcUAWorkarounds struct {
 	AdditionalValidStatusCodes []string `toml:"additional_valid_status_codes"`
 }
 
+type OpcUAAdditionalFields struct {
+	IncludeDataType bool `toml:"include_datatype"`
+}
```
Comment (Hipska) [inline plugins/common/opcua/client.go:23]: Not sure if this needs to be a new sub-struct.. (Just a comment, I'm open to other opinions)

**Review by Hipska — CHANGES_REQUESTED** [review]

File: plugins/inputs/opcua/sample.conf (hunk 1 above)
Comment (caallinson) [reply plugins/inputs/opcua/sample.conf:129]: ack

File: plugins/common/opcua/client.go (hunk 3 above)
Comment (caallinson) [reply plugins/common/opcua/client.go:23]: My thought here is that OPC-UA nodes have multiple attributes that can be read by a client including DataType, Description, namespace information, Quality (arguably). Putting them in a separate structure adds some organization as I anticipate future development will implement reading more of these attributes.

File: plugins/common/opcua/input/input_client.go [hunk 4]
```diff
@@ -421,6 +421,9 @@ func (o *OpcUAInputClient) MetricForNode(nodeIdx int) telegraf.Metric {
 
 	fields[nmm.Tag.FieldName] = o.LastReceivedData[nodeIdx].Value
 	fields["Quality"] = strings.TrimSpace(o.LastReceivedData[nodeIdx].Quality.Error())
+	if o.Config.OptionalFields.IncludeDataType {
+		fields["DataType"] = strings.Replace(o.LastReceivedData[nodeIdx].DataType.String(), "TypeID", "", 1)
+	}
```
[the remark this answers is not in the document]
Comment (Hipska) [reply plugins/common/opcua/input/input_client.go:426]: Thanks. Can you update the README with real output metric?

File: plugins/inputs/opcua_listener/opcua_listener_test.go [hunk 5]
```diff
@@ -148,6 +150,148 @@ func TestSubscribeClientIntegration(t *testing.T) {
...
+	container := testutil.Container{
+		Image:        "open62541/open62541",
+		ExposedPorts: []string{servicePort},
+		WaitingFor: wait.ForAll(
+			wait.ForListeningPort(nat.Port(servicePort)),
+			wait.ForLog("TCP network layer listening on opc.tcp://"),
+		),
+	}
+	err := container.Start()
+	require.NoError(t, err, "failed to start container")
```
Comment (Hipska) [inline plugins/inputs/opcua_listener/opcua_listener_test.go:167]:
```suggestion
	require.NoError(t, container.Start(), "failed to start container")
```

File: plugins/inputs/opcua_listener/opcua_listener_test.go [hunk 6]
```diff
@@ -148,6 +150,148 @@ func TestSubscribeClientIntegration(t *testing.T) {
...
+	require.NoError(t, err)
+
+	// give initial setup a couple extra attempts, as on CircleCI this can be
+	// attempted to soon
+	require.Eventually(t, func() bool {
+		return o.SetupOptions() == nil
+	}, 5*time.Second, 10*time.Millisecond)
+
+	err = o.Connect()
+	require.NoError(t, err, "Connection failed")
```
Comment (Hipska) [inline plugins/inputs/opcua_listener/opcua_listener_test.go:246]:
```suggestion
	require.NoError(t, o.Connect(), "Connection failed")
```

File: plugins/inputs/opcua_listener/opcua_listener_test.go [hunk 7]
```diff
@@ -148,6 +150,148 @@ func TestSubscribeClientIntegration(t *testing.T) {
...
+		case m := <-res:
+			for fieldName, fieldValue := range m.Fields() {
+				for _, tag := range testopctags {
+					if fieldName != tag.Name {
+						continue
+					}
+					if tag.Want == nil {
+						t.Errorf("Tag: %s has value: %v", tag.Name, fieldValue)
+						return
+					}
```
Comment (Hipska) [inline plugins/inputs/opcua_listener/opcua_listener_test.go:264]: This test is not clear. What is wrong with a tag having a certain value?

**Review by Hipska — CHANGES_REQUESTED** [review]
Good improvements, some minor comments left

File: plugins/inputs/opcua/opcua_test.go [hunk 8]
```diff
@@ -150,6 +152,97 @@ func TestReadClientIntegration(t *testing.T) {
...
+	container := testutil.Container{
+		Image:        "open62541/open62541",
+		ExposedPorts: []string{servicePort},
+		WaitingFor: wait.ForAll(
+			wait.ForListeningPort(nat.Port(servicePort)),
+			wait.ForLog("TCP network layer listening on opc.tcp://"),
+		),
+	}
+	err := container.Start()
+	require.NoError(t, err, "failed to start container")
```
[the remark this answers is not in the document]
Comment (caallinson) [reply plugins/inputs/opcua/opcua_test.go:169]: This is the common pattern in the codebase - I see 50 instances of it. Combine, or leave the same as the others?

File: plugins/inputs/opcua_listener/opcua_listener_test.go (hunk 7 above)
Comment (caallinson) [reply plugins/inputs/opcua_listener/opcua_listener_test.go:264]: That is from the base integration test - nil-valued tags are not expected to be sent to the client, so error if the server does send one.

File: plugins/common/opcua/client.go (hunk 3 above)
Comment (caallinson) [reply plugins/common/opcua/client.go:23]: Changed to a list

File: plugins/inputs/opcua_listener/opcua_listener_test.go (hunk 7 above)
Comment (Hipska) [reply plugins/inputs/opcua_listener/opcua_listener_test.go:263]: I meant; please make sure the error message explains the goal. Now it just says "Tag: x has value: y".

File: plugins/inputs/opcua/opcua_test.go [hunk 9]
```diff
@@ -150,6 +152,95 @@ func TestReadClientIntegration(t *testing.T) {
...
+	}
+
+	for _, tags := range testopctags {
+		readConfig.RootNodes = append(readConfig.RootNodes, MapOPCTag(tags))
+	}
+
+	client, err := readConfig.CreateReadClient(testutil.Logger{})
+	require.NoError(t, err)
+
+	require.NoError(t, client.Connect(), "Connect")
```
Comment (Hipska) [inline plugins/inputs/opcua/opcua_test.go:234]:
```suggestion
	require.NoError(t, client.Connect(), "Connection failed")
```
Or simply 
```suggestion
	require.NoError(t, client.Connect())
```

**Review by Hipska — APPROVED** [review]

File: plugins/common/opcua/client.go [hunk 10]
```diff
@@ -44,7 +44,8 @@ type OpcUAClientConfig struct {
 	ConnectTimeout config.Duration `toml:"connect_timeout"`
 	RequestTimeout config.Duration `toml:"request_timeout"`
 
-	Workarounds OpcUAWorkarounds `toml:"workarounds"`
+	OptionalFields []string         `toml:"optional_fields"`
```
Comment (srebhan) [inline plugins/common/opcua/client.go:47]: Could you please check if all parameters given for `optional_fields` are valid values? Currently, you silently ignore the option on misspellings like `optional_fields = ["datatype"]` or `optional_fields = ["DataTpye"]` which might be surprising to the user. I guess `func (o *OpcUAClientConfig) Validate() error` would be a good place to check.

**Review by srebhan — COMMENTED** [review]
Thanks for the PR @caallinson! One comment from my side...

File: plugins/common/opcua/client.go (hunk 10 above)
Comment (caallinson) [reply plugins/common/opcua/client.go:48]: Implemented

**Review by Hipska — CHANGES_REQUESTED** [review]

**Pull request merged**